pmd-ui/src/main/java/net/sourceforge/pmd/util/fxdesigner/MainDesignerController.java

```diff
@@ -0,0 +1,475 @@
+                Timeline timeline = new Timeline(new KeyFrame(Duration.millis(200), keyValue));
+                timeline.play();
+            }
+        });
+    }
+
+
+    public void shutdown() {
+        try {
+            XMLSettingsSaver saver = XMLSettingsSaver.forFile(DesignerUtil.getSettingsFile());
+            this.saveSettings(saver);
+            saver.save();
+        } catch (IOException ioe) {
+            // nevermind
+        }
+
+        sourceEditorController.shutdown(); // shutdown syntax highlighting
+        xpathPanelController.shutdown();
+    }
+
+
+    private void onRefreshASTClicked() {
+        sourceEditorController.refreshAST();
+        xpathPanelController.evaluateXPath(sourceEditorController.getCompilationUnit(),
+            sourceEditorController.getLanguageVersion());
+    }
+
+
+    /**
+     * Executed when the user selects a node in a treeView or listView.
+     */
+    public void onNodeItemSelected(Node selectedValue) {
+        nodeInfoPanelController.displayInfo(selectedValue);
+        sourceEditorController.clearNodeHighlight();
+        sourceEditorController.highlightNodePrimary(selectedValue);
+        sourceEditorController.focusNodeInTreeView(selectedValue);
+    }
+
+
+    public void onNameDeclarationSelected(NameDeclaration declaration) {
+        sourceEditorController.clearNodeHighlight();
+        sourceEditorController.highlightNodePrimary(declaration.getNode());
+        sourceEditorController.highlightNodesSecondary(declaration.getNode().getScope()
+                                                                  .getDeclarations()
+                                                                  .get(declaration)
+                                                                  .stream()
+                                                                  .map(NameOccurrence::getLocation)
+                                                                  .collect(Collectors.toList()));
+    }
+
+
+    private void showLicensePopup(ActionEvent event) {
+        Alert licenseAlert = new Alert(AlertType.INFORMATION);
+        licenseAlert.setWidth(500);
+        licenseAlert.setHeaderText("License");
+
+        ScrollPane scroll = new ScrollPane();
+        try {
+            scroll.setContent(new TextArea(IOUtils.toString(getClass().getResourceAsStream("LICENSE"))));
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        licenseAlert.getDialogPane().setContent(scroll);
+        licenseAlert.showAndWait();
+    }
+
+
+    private void onExportXPathToRuleClicked(Event event) throws IOException {
+        // doesn't work for some reason
+        ExportXPathWizardController wizard
+            = new ExportXPathWizardController(xpathPanelController.xpathExpressionProperty());
+
+        FXMLLoader loader = new FXMLLoader(getClass().getResource("fxml/xpath-export-wizard.fxml"));
+        loader.setControllerFactory(type -> {
```

Comment:
Instead of using a custom controller factory, it is also possible to request the controller later: After you have called `loader.load()`, you can call `loader.getController()` to retrieve the controller (might need to be casted), that has been created.
Then you could call a setter, to bind the xpathExpressionProperty.

Reply:
Nice idea, though I found out there's the even simpler `setController` method to which we can pass our controller without cast! Thanks though!